### ISSUE TITLE
Proposal: Add support for compiling CUDA with Clang

### ISSUE THREAD

**Issue by psalz** [illuhad/hipSYCL#22]
I’d love to see support for compiling CUDA using Clang (as detailed [here]( http://llvm.org/docs/CompileCudaWithLLVM.html)) added to `syclcc`. As far as I can tell this pretty much only has advantages over using `nvcc` (I could be wrong though, as I don’t have much CUDA experience).

For example, Clang will happily compile this program:

```c++
#include <CL/sycl.hpp>

int main() {
    cl::sycl::queue queue;

    queue.submit([](auto& cgh) {
        cgh.template parallel_for<class test>(cl::sycl::range<1>(10), [=](cl::sycl::item<1> item) {
            printf("Hello %d\n", item[0]);
        });
    });

    return 0;
}
```

As you know, generic lambas are among the things that `nvcc`’s (experimental!) extended lambda feature currently does not support at all. Don’t ask me why, but I actually need this.

As for feasibility: I added Clang as a third option into `syclcc` and successfully compiled both `libhipSYCL` itself (with very minor adjustments) as well as my application (which makes use of generic lambdas).

This would likely also allow us to omit the pruning step, as the call-graphs should be the same between the libTooling and Clang invocations.

Other advantages:
-	All (except for the `syclcc` specific) parameters can be simply passed to `clang++`, no juggling required.
-	You get Clang’s awesome error reporting.

In fact I think it might even be worth considering making Clang the default for compiling to CUDA, and dropping `nvcc` altogether. Anyone using hipSYCL has to have Clang installed anyway, so that is not a problem. However again, I’m not certain if there’s any downsides to this, like e.g. any features that aren’t supported. Dropping `nvcc` would allow us to use features that only Clang supports, like [overloading based on __device__ and __host__ attributes]( http://llvm.org/docs/CompileCudaWithLLVM.html#overloading-based-on-host-and-device-attributes). Although this might be moot as I’m not sure what `hcc`’s stance is on this.

**Comment by illuhad** [comment]
This is an issue where I can give you an answer very quickly: I'm very much in favor of adding clang-cuda support, and, in fact, had already thought about this myself :) If you already have it working, please file a pull request :)

I would still like to preserve nvcc support (even though it is a real pain). One argument for using hipSYCL is that people can use the latest hardware features and can get the same performance as with the proprietary nvcc compiler. In that way, there is zero reason for prefering CUDA over SYCL. This argument can only be made if nvcc is supported. clang (as far as I'm aware) doesn't support the latest cuda releases, produces sometimes slightly slower code and I suspect it may also take some time until new intrinsics end up in clang.
There are also other problems with nvcc, for example it has restrictions using local types in kernels which in particular affects SYCL parallel STL code. I was contemplating dealing with nvcc extended lambda peculiarities in the future by rewriting them into function objects, which generally don't suffer from such restrictions..

**Comment by psalz** [comment]
I've actually been doing my testing with a recent Clang 8 nightly (since we have a bunch of 2070s around here that need CUDA 10), so I can confirm that it works! According to the LLVM website their 8.0.0 release is scheduled for February 27th, so that makes it about half a year after CUDA 10. Not too bad!

I've actually also spent half a day last week fiddling around with libTooling to see if rewriting lambdas into functors would be feasible, but seeing all of the issues that came up with the pruning already I'm sure glad that I can just use Clang 8 instead ;-).

I also very much agree with the idea of making Clang the default for CUDA. However I should mention that I did run into a compilation error on one project that looked like maybe a missing intrinsic or something, but I haven't investigated it thoroughly yet. I will look into this some more and also try and cook up a PR hopefully tomorrow!

**Comment by illuhad** [comment]
hm, even if there are a couple of intrinsics missing I don't think that's really a problem for us:
* In order to implement SYCL we actually only need very few intrinsics (the basic stuff like `__syncthreads()` and atomics which clang certainly supports). So, the average user who just wants to write SYCL code won't notice if clang supports all intrinsics that nvcc does.
* The power user who wants to make use of hipSYCL's capability of using low-level optimizations with CUDA intrinsics inside SYCL code can always have such optimized kernels in a separate cpp file that is compiled with `syclcc --hipsycl-platform=nvcc` or compile his entire project with nvcc (potentially only selectively rewriting includes), if regular clang doesn't work.

While catering to the needs of such power users is a core feature of hipSYCL, I think we first should get the fundamentals right for the average user, and clang helps a great deal here.

If clang keeps up with being half a year behind CUDA that's very promising, although we shouldn't forget that it may take significantly longer for new versions of clang to actually land in Linux distributions. I would prefer if the average user wouldn't have to put up with compiling clang - I'm also at the moment trying to compile llvm/clang 8 now, and it's been going for hours ;) But on the other hand, power users who have the newest hardware and need the latest version of CUDA may have enough computing power to compile llvm very quickly ;)

Looking forward to merging your PR! :)

**Comment by psalz** [comment]
I forgot to report my findings regarding this potential missing intrinsic: It was about `__builtin_ia32_cvtdq2ps` (so for x86), and it was again within a header that doesn't contain any device code. Not inlining the header fixes it (another one of those applications for #20).

**Comment by illuhad** [comment]
that seems weird, why would clang's cuda compiler not know about some x86 intrinsic? It can't really be because of some compiler misidentification since source transformation and actual compilation are all handled by clang?

**Comment by illuhad** [comment]
Implemented in PR #27.
